src/snapshot.ts

```diff
@@ -40,6 +50,10 @@ const snapshot = (value: any): unknown => {
           modelSnapshot[name] = snapshot((value as any)[name]);
         }
       }
+      if (isReadOnly) {
+        cache.set(value, modelSnapshot);
```

Comment:
👍 game for the `WeakMap` approach. We could also store a `WeakRef` on some symbol key too. Not sure if that'd perform better than the `WeakMap` lookup.